### PR TITLE
Prepare SMT comparison for counterexamples

### PR DESCRIPTION
**Pull request by leonardoalt**
Needed by https://github.com/ethereum/solidity/pull/10477

### CODE REVIEW

File: test/smtcallback.js [hunk 1]
```diff
@@ -146,6 +147,11 @@ tape('SMTCheckerCallback', function (t) {
         continue;
       }
 
+      if (source.includes('// SMTIgnoreCex: yes')) {
+        st.skip('Test ignores counterexamples.');
+        continue;
+      }
```
Comment (ekpyron) [inline test/smtcallback.js:153]: So you don't bother to just ignore the counterexamples here, but skip the entire test :-)?
You *could* just cut the expectations short after "happens here" in those cases, couldn't you?

**Review by ekpyron — APPROVED** [review]
Not sure how @axic feels about the coverage decrease ;-). And we could also do something more elaborate here IIUC... but it'd be fine with me, so tentative approval...

File: test/smtcallback.js (hunk 1 above)
Comment (leonardoalt) [reply test/smtcallback.js:153]: Yes, I found that easier. It's only a handful of tests and the counterexamples are tested on the rest so, no harm I'd say. Just trying to keep it simple. On the other hand, I could also do that, yes.

Comment (leonardoalt) [reply test/smtcallback.js]: Updated

**Comment by ekpyron** [comment]
Tests failing.

**Comment by leonardoalt** [comment]
> Tests failing.

No way

**Comment by leonardoalt** [comment]
We should also wait for https://app.circleci.com/pipelines/github/ethereum/solidity/11293/workflows/3cdf4886-1d63-42b0-b84e-f59e8ead7d6f/jobs/541919
that's the `t_ems_solcjs` from the Solidity PR using this PR

**Review by ekpyron — APPROVED** [review]
So yeah... still fine with me... coverage is still an open question...

**Comment by leonardoalt** [comment]
Added a test directly here, let's see if that affects coverage. Otherwise, the Solidity `t_ems_solcjs` do test the added lines.

**Comment by leonardoalt** [comment]
Didn't change it, I think we should merge as is.

**Comment by ekpyron** [comment]
> Didn't change it, I think we should merge as is.

Since coverage is set to be required here, that'd need @axic's approval and doing then, I guess :-).

**Comment by leonardoalt** [comment]
Ah, right :p

**Comment by axic** [comment]
> coverage is still an open question...

Coverage has been steadily decreasing since I made the effort to reach close to 100% coverage a year or two ago 😅

**Comment by axic** [comment]
Which PR introduced SMTIgnoreCex ?

**Comment by leonardoalt** [comment]
@axic: https://github.com/ethereum/solidity/pull/10477 which is actually waiting for this. Right now it points to this branch, so `t_ems_solcjs` pass there

**Comment by axic** [comment]
That PR only seems to be using that field, it did not introduce it?

**Comment by leonardoalt** [comment]
@axic it also introduced it in `test/libsolidity/SMTCheckerTest.cpp`:
https://github.com/ethereum/solidity/pull/10477/files#r538771272

File: test/smtcallback.js [hunk 2]
```diff
@@ -146,6 +151,11 @@ tape('SMTCheckerCallback', function (t) {
         continue;
       }
 
+      let ignoreCounterexamples = false;
```
Comment (axic) [inline test/smtcallback.js:154]: If you use `ignoreCex` above and in the test sources, perhaps use the same here?

File: test/smtcallback.js [hunk 3]
```diff
@@ -34,7 +34,12 @@ function expectErrors (errors, expectations) {
     if (errors[i].includes('Error trying to invoke SMT solver') || expectations[i].includes('Error trying to invoke SMT solver')) {
       continue;
     }
-    expectations[i] = expectations[i].replace('happens here.', 'happens here');
+    expectations[i] = expectations[i].replace('happens here.', 'happens here').split('\\n').join('\n');
+    errors[i] = errors[i].replace('happens here.', 'happens here').split('\\n').join('\n');
+    if (ignoreCex) {
+      expectations[i] = expectations[i].split('\n')[0];
```
Comment (axic) [inline test/smtcallback.js:40]: I don't fully understand why is it done this way. Why not just cut off anything after `Counterexamples:` ?

File: test/smtcallback.js (hunk 2 above)
Comment (leonardoalt) [reply test/smtcallback.js:154]: I wanted to, but in the `test` object it's also called `ignoreCex`, I wasn't sure about shadowing in the assignment below

File: test/smtcallback.js (hunk 3 above)
Comment (leonardoalt) [reply test/smtcallback.js]: Updated

File: test/smtcallback.js [hunk 4]
```diff
@@ -34,7 +34,12 @@ function expectErrors (errors, expectations) {
     if (errors[i].includes('Error trying to invoke SMT solver') || expectations[i].includes('Error trying to invoke SMT solver')) {
       continue;
     }
-    expectations[i] = expectations[i].replace('happens here.', 'happens here');
+    expectations[i] = expectations[i].replace('happens here.', 'happens here').split('\\n').join('\n');
+    errors[i] = errors[i].replace('happens here.', 'happens here').split('\\n').join('\n');
+    if (ignoreCex) {
+      expectations[i] = expectations[i].split('\nCounterexample')[0];
```
Comment (axic) [inline test/smtcallback.js:40]: If we do this, do we actually need to do that split outside of this branch?

Comment (leonardoalt) [reply test/smtcallback.js:40]: I think so. Somehow, weirdly it reads one of them as `\\n` and the other as `\n`

Comment (leonardoalt) [reply test/smtcallback.js:40]: So even if we split at `Counterexample`, `\\n` won't match `\n` when before `Counterexample`

Comment (leonardoalt) [reply test/smtcallback.js:40]: When I print `test.expectations` and `test.errors`:
```
[
  '// Warning 2529: (82-89): CHC: Empty array "pop" happens here.\\nCounterexample:\\na = []\\n\\n\\n\\nTransaction trace:\\nconstructor()\\nState: a = []\\nf()'
]
[
  'CHC: Empty array "pop" happens here.\n' +
    'Counterexample:\n' +
    'a = []\n' +
    '\n' +
    '\n' +
    '\n' +
    'Transaction trace:\n' +
    'constructor()\n' +
    'State: a = []\n' +
    'f()'
]
```

File: test/smtcallback.js [hunk 5]
```diff
@@ -34,8 +34,19 @@ function expectErrors (errors, expectations) {
...
       continue;
     }
-    expectations[i] = expectations[i].replace('happens here.', 'happens here');
-    if (!errors[i].includes(expectations[i])) {
+    // Expectations containing counterexamples might have many '\n' in a single line.
+    // When these are read, they come as '\\n', whereas the actual error from the compile has '\n'.
+    // Therefore we need to replace '\\n' by '\n' in the expectations.
+    // Function `replace` only replaces the first occurrence, and `replaceAll` is not standard yet.
+    // Replace all '\\n' by '\n' via split & join.
+    expectations[i] = expectations[i].split('\\n').join('\n');
```
Comment (leonardoalt) [inline test/smtcallback.js:42]: This is needed here for the case where we actually want to compare the counterexamples

File: test/smtcallback.js [hunk 6]
```diff
@@ -146,6 +157,11 @@ tape('SMTCheckerCallback', function (t) {
         continue;
       }
 
+      let ignoreCex = false;
+      if (source.includes('// SMTIgnoreCex: yes')) {
```
Comment (axic) [inline test/smtcallback.js:161]: Couldn't actually just do `let ignoreCex =  source.includes('// SMTIgnoreCex: yes')` ?

Comment (leonardoalt) [reply test/smtcallback.js:161]: Probably

File: test/smtcallback.js [hunk 7]
```diff
@@ -34,8 +34,19 @@ function expectErrors (errors, expectations) {
     if (errors[i].includes('Error trying to invoke SMT solver') || expectations[i].includes('Error trying to invoke SMT solver')) {
       continue;
     }
-    expectations[i] = expectations[i].replace('happens here.', 'happens here');
-    if (!errors[i].includes(expectations[i])) {
+    // Expectations containing counterexamples might have many '\n' in a single line.
+    // When these are read, they come as '\\n', whereas the actual error from the compile has '\n'.
```
Comment (axic) [inline test/smtcallback.js:38]:
```suggestion
    // When these are read, they come as '\\n', whereas the actual error from the compiler has '\n'.
```

Comment (axic) [inline test/smtcallback.js:38]:
```suggestion
    // These are stored escaped in test format (as '\\n'), whereas the actual error from the compile has '\n'.
```

**Review by axic — APPROVED** [review]

**Pull request merged**